release/models/route-programming/openconfig-route-programming-state.yang

```diff
@@ -0,0 +1,186 @@
+        description
+          "Surrounding container for the list of routes that fail programming.";
+
+        list failed {
+          key "prefix";
+
+          description
+            "A prefix that was attempted to the programmed into hardware, but the
+            programming operation failed.";
+
+          leaf prefix {
+            type leafref {
+              path "../state/prefix";
+            }
+            description
+              "Reference to the prefix that keys the failed list.";
+          }
+
+          container state {
+            description
+              "Operational state parameters relating to a failed programming
+              operation.";
+            uses route-state-common;
+          }
+        }
+      }
+
+      container stale-routes {
+        description
+          "Surrounding container for the list of routes that are currently in
+          a stale state.";
+
+        list stale {
+          key "prefix";
+
+          description
+            "A prefix that is currently installed in hardware, but a subsequent
+            operation to update its programming failed - such that the entry in
+            hardware is stale.";
+
+          leaf prefix {
+            type leafref {
+              path "../state/prefix";
+            }
+            description
+              "Reference to the prefix that keys the stale list.";
+          }
+
+          container state {
+            description
+              "Operational state parameters relating to a stale route.";
+            uses route-state-common;
+          }
+        }
+      }
+
+      container drop-routes {
+        description
+          "Surrounding container for the list of routes that are currently in
+          a drop state.";
+
+        list drop {
+          key "prefix";
+
+          description
+            "A prefix that is currently installed in hardware but with an explicit
+            instruction that it should discard packets that are destined towards
+            it.";
+
+          leaf prefix {
+            type leafref {
+              path "../state/prefix";
+            }
+            description
+              "Reference to the prefix that keys the drop list.";
+          }
+
+          container state {
+            description
+              "Operational state parameters relating to a drop route.";
+            uses route-state-common;
+          }
+        }
+      }
+    }
+  }
+
+ grouping route-state-common {
+    description
+      "Common state parameters that correspond to a particular route
+      type.";
+
+    leaf prefix {
+      type oc-inet:ip-prefix;
+      description
+        "The IPv4 or IPv6 prefix that the route state corresponds to.";
+    }
+
+    leaf time {
+      type oc-types:timeticks64;
+      description
+        "Time that the route was programmed, or the programming attempt was made
+        expressed as nanoseconds since the Unix epoch.";
+    }
+
+    leaf reason {
```

Comment:
As mentioned in some general observations, a programming error could occur on one component in a multi-component system but there is currently no signal into such.  While this reason gives a generic code to the client, does it make sense to include an opaque leaf to signal additional information helpful to the client?  The other option is fully modeled structures to represent components.

Reply:
Do you mean for example, in a two ASIC system, route programming could succeed on one ASIC and fail on a second?    Are you suggesting something like:
```
leaf dest-component {
  type leafref {
    path "/oc-platform:components/oc-platform:component";
  }
  description
    "The destination component for the route programming";
 }
```

Reply:
Correct - a reference to a component at least gives an indication as to where the failure was sourced.  An additional leafref as you have above in `route-state-common` seems appropriate for such

Reply:
Just to confirm, if the publishing logic doesn't quite know the component, the expectation would be for it to just skip populating this leaf correct? ( As I understand non-key leaf attributes are by default optional )

Reply:
You would have the choice of either
* not populating the leaf
* deviating the leaf altogether

The first case allows for populating when you know and not when you do not.  The latter covers non-support altogether

General rule of thumb across r/o data being optional is what you describe however this should be documented at a minimum for OC gRPC services wrt. client/server expectations